src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/Tensor.cs

```diff
@@ -1,751 +1,5624 @@
+                scoped Span<nint> newStrides = TensorOperation.RentedBuffer.CreateUninitialized(tensor.Rank, out TensorOperation.RentedBuffer<nint> stridesRentedBuffer);
+                scoped Span<int> newLinearOrder = TensorOperation.RentedBuffer.CreateUninitialized(tensor.Rank, out TensorOperation.RentedBuffer<int> linearOrderRentedBuffer);
+
+                Tensor<T> outTensor;
+
+                if (dimensions.IsEmpty)
+                {
+                    for (int i = 0; i < tensor.Rank; i++)
+                    {
+                        newLengths[i] = tensor.Lengths[tensor.Rank - 1 - i];
+                        newStrides[i] = tensor.Strides[tensor.Rank - 1 - i];
+                        newLinearOrder[i] = tensor._shape.LinearRankOrder[tensor.Rank - 1 - i];
+                    }
+                }
+                else
+                {
+                    if (dimensions.Length != tensor.Lengths.Length)
+                        ThrowHelper.ThrowArgument_PermuteAxisOrder();
+
+                    for (int i = 0; i < dimensions.Length; i++)
+                    {
+                        newLengths[i] = tensor.Lengths[dimensions[i]];
+                        newStrides[i] = tensor.Strides[dimensions[i]];
+                        newLinearOrder[i] = tensor._shape.LinearRankOrder[dimensions[i]];
+                    }
+                }
+                outTensor = new Tensor<T>(tensor._values, tensor._start, newLengths, newStrides, newLinearOrder);
+
+                lengthsRentedBuffer.Dispose();
+                stridesRentedBuffer.Dispose();
+                linearOrderRentedBuffer.Dispose();
+
+                return outTensor;
+            }
+        }
+        #endregion
+
+        #region Reshape
+        /// <summary>
+        /// Reshapes the <paramref name="tensor"/> tensor to the specified <paramref name="lengths"/>. If one of the lengths is -1, it will be calculated automatically.
+        /// Does not change the length of the underlying memory nor does it allocate new memory. If the new shape is not compatible with the old shape,
+        /// an exception is thrown.
+        /// </summary>
+        /// <param name="tensor"><see cref="Tensor{T}"/> you want to reshape.</param>
+        /// <param name="lengths"><see cref="ReadOnlySpan{T}"/> with the new dimensions.</param>
+        public static Tensor<T> Reshape<T>(this Tensor<T> tensor, ReadOnlySpan<nint> lengths)
+        {
+            if (tensor.Lengths.SequenceEqual(lengths))
+                return tensor;
+
+            if (!tensor.IsContiguousAndDense && !tensor.Strides.Contains(0))
+            {
+                ThrowHelper.ThrowArgument_CannotReshapeNonContiguousOrDense();
+            }
+
+            nint[] newLengths = lengths.ToArray();
+            // Calculate wildcard info.
+            if (lengths.Contains(-1))
```

Comment:
nit: doing` lengths.IndexOf(-1)` better to avoid calling it again later?

Reply:
Good idea. fixed locally for followup pr.